File: local-scanner/python/network_scanner.py

#!/usr/bin/env python3
import json
import sys
import socket
import subprocess
import platform
import re
from datetime import datetime

def get_os():
    """Detect the operating system."""
    return platform.system()

def scan_network_devices():
    """Scan for devices on the network and return a list of them."""
    os_name = get_os()
    devices = []
    
    try:
        if os_name == "Linux" or os_name == "Darwin":  # Linux or macOS
            # Use ARP to get network devices
            arp_output = subprocess.check_output(["arp", "-a"]).decode("utf-8")
            lines = arp_output.split("\n")
            
            for line in lines:
                if not line.strip():
                    continue
                    
                # Parse ARP output
                match = re.search(r"\((\d+\.\d+\.\d+\.\d+)\) at ([0-9a-f:]+)", line, re.IGNORECASE)
                if match:
                    ip = match.group(1)
                    mac = match.group(2)
                    
                    # Try to get hostname
                    try:
                        hostname = socket.gethostbyaddr(ip)[0]
                    except:
                        hostname = "Unknown"
                    
                    # Determine device type based on MAC address (simplified)
                    device_type = "Unknown"
                    if mac.startswith("00:0C:29") or mac.startswith("00:50:56") or mac.startswith("00:05:69"):
                        device_type = "Virtual Machine"
                    elif mac.startswith("00:1A:11") or mac.startswith("00:25:00"):
                        device_type = "Smartphone"
                    elif mac.startswith("B8:27:EB") or mac.startswith("DC:A6:32"):
                        device_type = "Raspberry Pi"
                    
                    devices.append({
                        "ip": ip,
                        "mac": mac,
                        "name": hostname,
                        "type": device_type,
                        "status": "Online"
                    })
                    
        elif os_name == "Windows":
            # For Windows, use the 'arp -a' command
            arp_output = subprocess.check_output(["arp", "-a"], shell=True).decode("utf-8")
            lines = arp_output.split("\n")
            
            for line in lines:
                if "dynamic" in line.lower():
                    parts = line.split()
                    if len(parts) >= 2:
                        ip = parts[0].strip()
                        mac = parts[1].strip()
                        
                        # Try to get hostname
                        try:
                            hostname = socket.gethostbyaddr(ip)[0]
                        except:
                            hostname = "Unknown"
                        
                        devices.append({
                            "ip": ip,
                            "mac": mac,
                            "name": hostname,
                            "type": "Unknown",
                            "status": "Online"
                        })
        
        # Additional scan using nmap if available (optional)
        try:
            nmap_output = subprocess.check_output(["nmap", "-sP", "192.168.1.0/24"], stderr=subprocess.DEVNULL).decode("utf-8")
            # Process nmap output if needed
        except:
            pass  # nmap not available or failed
            
    except Exception as e:
        print(f"Error during scan: {str(e)}", file=sys.stderr)
    
    return devices

def get_device_details(ip):
    """Get detailed information about a specific device."""
    device_info = {
        "ip": ip,
        "mac": "",
        "name": "Unknown",
        "type": "Unknown",
        "ports": [],
        "os": "Unknown",
        "status": "Unknown"
    }
    
    try:
        # Get MAC address
        if get_os() in ["Linux", "Darwin"]:
            arp_output = subprocess.check_output(["arp", "-n", ip]).decode("utf-8")
            mac_match = re.search(r"([0-9a-f]{1,2}:[0-9a-f]{1,2}:[0-9a-f]{1,2}:[0-9a-f]{1,2}:[0-9a-f]{1,2}:[0-9a-f]{1,2})", 
                                 arp_output, re.IGNORECASE)
            if mac_match:
                device_info["mac"] = mac_match.group(1)
        
        # Try to get hostname
        try:
            device_info["name"] = socket.gethostbyaddr(ip)[0]
        except:
            pass
        
        # Check if device is up
        ping_param = "-n" if get_os() == "Windows" else "-c"
        ping_output = subprocess.call(["ping", ping_param, "1", "-w", "1", ip], 
                                     stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        device_info["status"] = "Online" if ping_output == 0 else "Offline"
        
        # Try to determine OS (simplified)
        if device_info["status"] == "Online":
            try:
                ttl_output = subprocess.check_output(["ping", ping_param, "1", ip]).decode("utf-8")
                ttl_match = re.search(r"ttl=(\d+)", ttl_output, re.IGNORECASE)
                if ttl_match:
                    ttl = int(ttl_match.group(1))
                    if ttl <= 64:
                        device_info["os"] = "Linux/Unix"
                    elif ttl <= 128:
                        device_info["os"] = "Windows"
                    else:
                        device_info["os"] = "Network Equipment"
            except:
                pass
    
    except Exception as e:
        print(f"Error getting details for {ip}: {str(e)}", file=sys.stderr)
    
    return device_info

if __name__ == "__main__":
    # Simple command processing
    if len(sys.argv) < 2:
        print("Usage: python3 network_scanner.py [command] [parameters]")
        sys.exit(1)
    
    command = sys.argv[1]
    
    if command == "scan":
        devices = scan_network_devices()
        print(json.dumps(devices))
    elif command == "device" and len(sys.argv) >= 3:
        ip_address = sys.argv[2]
        details = get_device_details(ip_address)
        print(json.dumps(details))
    else:
        print(f"Unknown command: {command}")
        sys.exit(1)

File: local-scanner/networkScanner.ts

import { execSync } from 'child_process';
import { spawn } from 'child_process';
import { existsSync } from 'fs';
import path from 'path';
import { NetworkDevice } from '../src/hooks/network/types';

// Helper to determine the correct path to the Python script
const getPythonScriptPath = () => {
  const relativePath = 'python/network_scanner.py';
  const scriptPath = path.join(__dirname, relativePath);
  return scriptPath;
};

// Helper to check if Python is available
const isPythonAvailable = (): boolean => {
  try {
    // Check for Python 3 first
    execSync('python3 --version', { stdio: 'ignore' });
    return true;
  } catch (e) {
    try {
      // Fallback to python command
      execSync('python --version', { stdio: 'ignore' });
      return true;
    } catch (e) {
      console.log('Python is not available on this system');
      return false;
    }
  }
};

// Helper to run Python script with proper command
const runPythonScript = (args: string[]): Promise<string> => {
  return new Promise((resolve, reject) => {
    const scriptPath = getPythonScriptPath();
    
    // Check if the script exists
    if (!existsSync(scriptPath)) {
      reject(new Error(`Python script not found: ${scriptPath}`));
      return;
    }

    // Determine which python command to use
    const pythonCommand = (() => {
      try {
        execSync('python3 --version', { stdio: 'ignore' });
        return 'python3';
      } catch (e) {
        return 'python';
      }
    })();

    console.log(`Running Python script: ${pythonCommand} ${scriptPath} ${args.join(' ')}`);
    
    const proc = spawn(pythonCommand, [scriptPath, ...args]);
    
    let stdout = '';
    let stderr = '';
    
    proc.stdout.on('data', (data) => {
      stdout += data.toString();
    });
    
    proc.stderr.on('data', (data) => {
      stderr += data.toString();
      console.error(`Python stderr: ${stderr}`);
    });
    
    proc.on('close', (code) => {
      if (code !== 0) {
        if (stderr.includes('netifaces')) {
          console.warn('Python netifaces module not available. Using fallback methods.');
          // Provide a simplified response when netifaces is not available
          if (args[0] === 'scan') {
            resolve(JSON.stringify([])); // Empty device list
          } else if (args[0] === 'device') {
            resolve(JSON.stringify({
              ip: args[1],
              mac: 'Unknown',
              name: 'Unknown Device',
              type: 'Unknown',
              status: 'Unknown',
              id: Math.random().toString(36).substring(2, 10)
            }));
          } else if (args[0] === 'status') {
            resolve(JSON.stringify({
              pythonAvailable: true,
              modules: {
                scapy: false,
                nmap: false,
                netifaces: false,
                psutil: false
              },
              os: process.platform
            }));
          }
        } else {
          reject(new Error(`Python script failed with code ${code}: ${stderr}`));
        }
      } else {
        resolve(stdout);
      }
    });
    
    proc.on('error', (err) => {
      console.error(`Python process error: ${err.message}`);
      reject(err);
    });
  });
};

export const networkScanner = {
  isPythonAvailable: isPythonAvailable(),
  runPythonScript, // Export for use in other files
  
  scan: async (): Promise<NetworkDevice[]> => {
    try {
      // Try to use Python scanner first
      if (isPythonAvailable()) {
        console.log('Using Python network scanner');
        const output = await runPythonScript(['scan']);
        try {
          return JSON.parse(output) as NetworkDevice[];
        } catch (e) {
          console.error('Failed to parse Python scanner output:', e);
          console.log('Python output:', output);
          throw e; // Let the next block handle fallback
        }
      } 
      
      // Fallback to original method
      console.log('Using fallback network scanner (ARP)');
      try {
        const output = execSync('arp -a').toString();
        const devices = output.split('\n')
          .filter(line => line.includes('('))
          .map(line => {
            const [ip, mac] = line.split(/\s+/);
            return {
              ip,
              mac,
              name: 'Unknown Device',
              type: 'Unknown',
              status: 'Online',
              id: Math.random().toString(36).substring(2, 10)
            };
          });
        
        return devices;
      } catch (arpError) {
        console.error('ARP command failed:', arpError);
        // Return empty array if both Python and ARP fail
        return [];
      }
    } catch (error) {
      console.error('Network scan error:', error);
      return [];
    }
  },
  
  getDeviceDetails: async (ipAddress: string): Promise<NetworkDevice | null> => {
    try {
      if (isPythonAvailable()) {
        console.log(`Getting device details for IP: ${ipAddress} using Python`);
        const output = await runPythonScript(['device', ipAddress]);
        try {
          const device = JSON.parse(output) as NetworkDevice;
          if (!device.id) {
            device.id = Math.random().toString(36).substring(2, 10);
          }
          return device;
        } catch (e) {
          console.error(`Failed to parse device details for ${ipAddress}:`, e);
          console.log('Python output:', output);
          return null;
        }
      }
      
      console.log(`Python unavailable, using fallback for device ${ipAddress}`);
      return {
        ip: ipAddress,
        mac: 'Unknown',
        name: 'Unknown Device',
        type: 'Unknown',
        status: 'Unknown',
        id: Math.random().toString(36).substring(2, 10)
      };
    } catch (error) {
      console.error(`Error getting device details for ${ipAddress}:`, error);
      return null;
    }
  }
};
